File: game-engine/src/main/java/org/luxons/sevenwonders/game/moves/DiscardMove.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.cards.Card;

public class DiscardMove extends CardFromHandMove {

    DiscardMove(int playerIndex, Card card, PlayerMove move) {
        super(playerIndex, card, move);
    }

    @Override
    public void place(Table table, List<Card> discardedCards, Settings settings) {
        discardedCards.add(getCard());
    }

    @Override
    public void activate(Table table, List<Card> discardedCards, Settings settings) {
        Board board = table.getBoard(getPlayerIndex());
        board.addGold(settings.getDiscardedCardGold());
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/DiscardMove.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.Settings
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.api.Table
import org.luxons.sevenwonders.game.boards.Board
import org.luxons.sevenwonders.game.cards.Card

class DiscardMove internal constructor(playerIndex: Int, card: Card, move: PlayerMove) :
    CardFromHandMove(playerIndex, card, move) {

    override fun place(table: Table, discardedCards: MutableList<Card>, settings: Settings) {
        discardedCards.add(card)
    }

    override fun activate(table: Table, discardedCards: List<Card>, settings: Settings) {
        val board = table.getBoard(playerIndex)
        board.addGold(settings.discardedCardGold)
    }
}
